pkg/operator/controllers/azurensg/subnetnsg.go

```diff
@@ -0,0 +1,139 @@
+	if err != nil {
+		return err
+	}
+	// Grab azure-credentials from secret
+	credentials, err := clusterauthorizer.AzCredentials(ctx, r.kubernetescli)
+	if err != nil {
+		return err
+	}
+	resource, err := azure.ParseResourceID(instance.Spec.ResourceID)
+	if err != nil {
+		return err
+	}
+	// create service principal token from azure-credentials
+	token, err := aad.GetToken(ctx, r.log, string(credentials.ClientID), string(credentials.ClientSecret), string(credentials.TenantID), azEnv.ActiveDirectoryEndpoint, azEnv.ResourceManagerEndpoint)
+	if err != nil {
+		return err
+	}
+	// create refreshable authorizer from token
+	authorizer, err := clusterauthorizer.NewAuthorizer(token)
+	if err != nil {
+		return err
+	}
+	subnetsClient := network.NewSubnetsClient(&azEnv, resource.SubscriptionID, authorizer)
+	// the main logic starts here
+	subnets, masterResourceGroup, err := r.getSubnets(ctx)
+	if err != nil {
+		return err
+	}
+	for ws := range subnets {
+		err = r.fixSubnet(ctx, api.ArchitectureVersion(instance.Spec.ArchitectureVersion), resource.SubscriptionID, ws.vnetName, ws.subnetName, ws.resourceGroup, masterResourceGroup, instance.Spec.InfraID, subnets[ws], subnetsClient)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *AzureNSGReconciler) fixSubnet(ctx context.Context, architectureVersion api.ArchitectureVersion, subscriptionID, vnetName, subnetName, vnetResourceGroup, resourcesResourceGroup, infraID string, isWorkerSubnet bool, subnetsClient network.SubnetsClient) error {
+	fullSubnet, err := subnetsClient.Get(ctx, vnetResourceGroup, vnetName, subnetName, "")
+	if err != nil {
+		return err
+	}
+	if fullSubnet.SubnetPropertiesFormat == nil {
+		return fmt.Errorf("received nil SubnetProperties when trying to Get subnet %s/%s in resource group %s", vnetName, subnetName, vnetResourceGroup)
+	}
+	if fullSubnet.SubnetPropertiesFormat.NetworkSecurityGroup == nil {
+		return fmt.Errorf("received nil NSG when trying to Get subnet %s/%s in resource group %s", vnetName, subnetName, vnetResourceGroup)
+	}
+	correctNSGResourceID, err := subnet.NetworkSecurityGroupIDExpanded(architectureVersion, resourcesResourceGroup, subnetName, infraID, isWorkerSubnet)
+	if err != nil {
+		return err
+	}
+	correctNSGResourceID = fmt.Sprintf("/subscriptions/%s/resourceGroups/%s", subscriptionID, correctNSGResourceID)
+
+	if !strings.EqualFold(*fullSubnet.NetworkSecurityGroup.ID, correctNSGResourceID) {
+		r.log.Infof("Fixing NSG from %s to %s", *fullSubnet.NetworkSecurityGroup.ID, correctNSGResourceID)
+		// NSG doesn't match - fixing
+		fullSubnet.NetworkSecurityGroup = &mgmtnetwork.SecurityGroup{ID: &correctNSGResourceID}
+		err = subnetsClient.CreateOrUpdateAndWait(ctx, vnetResourceGroup, vnetName, subnetName, fullSubnet)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *AzureNSGReconciler) getSubnets(ctx context.Context) (map[subnetDescriptor]bool, string, error) {
```

Comment:
This function is so complicated in asks for unit tests :/

Reply:
Just realised. Do we care about machine roles introduced by customers (e.g. infra)?

Reply:
We do, This is why I think making it abstract and saying - one map and all used subsets (might be more than 2). And we just identify masters one, and get others.

Reply:
In the current version of PR we only get master machines and worker machines.

So I guess we need to look into whether or not we can do `LabelSelector` to be "master" and "not master". Should be [possible](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#set-based-requirement)..

Reply:
Changed label selectors to "machine.openshift.io/cluster-api-machine-role!=master" and "machine.openshift.io/cluster-api-machine-role=master". Writing a unit test for the getSubnets function.

Reply:
Added unit tests for the getSubnets function, but I'll have to think about more failure modes. So far it tests the main code path + 2 cases of missing data.